### ISSUE TITLE
cluster terminates instances that have jobs running

### ISSUE THREAD

**Issue by msherman13** [aws/aws-parallelcluster#1145]
**Environment:**
 - AWS ParallelCluster v2.3.1
 - OS: centos7
 - Scheduler: SGE
 - Master instance type: t2.micro
 - Compute instance type: c4.8xlarge

**Bug description and how to reproduce:**
When I submit a large number of jobs to the cluster, and then some time later submit another set of jobs (after the initial jobs are already running), some instances with running jobs get terminated. The result is that the jobs which were running get hung in qstat and the node cannot be removed from qhost unless I forcefully delete the jobs.

**Additional context:**
In sqswatcher, I see that the instance in question is created:
2019-06-16 19:16:10,701 INFO [sqswatcher:_parse_sqs_messages] Processing COMPUTE_READY event for instance i-053275a57b3cc9fb4

and then when I submit the additional jobs later, I see this in sqswatcher:
2019-06-16 19:25:53,133 INFO [sqswatcher:_parse_sqs_messages] Processing EC2_INSTANCE_TERMINATE event for instance i-053275a57b3cc9fb4
2019-06-16 19:25:53,141 INFO [sqswatcher:_parse_sqs_messages] Processing EC2_INSTANCE_TERMINATE event for instance i-001408cd3b1bcb613
2019-06-16 19:25:53,147 ERROR [sqswatcher:_process_instance_terminate_event] Instance i-001408cd3b1bcb613 not found in the database.

when I look at the instance status from the aws cli, I see that the instance was terminated by user:
                    "StateReason": {
                        "Message": "Client.UserInitiatedShutdown: User initiated shutdown", 
                        "Code": "Client.UserInitiatedShutdown"
                    }, 
                    "State": {
                        "Code": 48,  
                        "Name": "terminated"
                    }, 
                    "EbsOptimized": false, 
                    "LaunchTime": "2019-06-16T19:10:48.000Z", 
                    "ProductCodes": [],  
                    "CpuOptions": {
                        "CoreCount": 18,  
                        "ThreadsPerCore": 2
                    }, 
                    "StateTransitionReason": "User initiated (2019-06-16 19:24:55 GMT)", 
                    "InstanceId": "i-053275a57b3cc9fb4", 

Note that I am using spot instances, but it doesn't seem like this is the result of aws pulling the instance. And I have the bid price set to the default, which I believe is the price of the on-demand instance.

**Comment by demartinofra** [comment]
Hi @msherman13 

Thank you for reporting this.
Without the nodewatcher logs of the terminated instance it is hard to tell what's going on but here are some potential root causes and relative fixes:
- the t2.micro instance used as Master node might be the culprit. I'm afraid that the tiny instance cannot handle the cluster load (especially when the size of the cluster grows) and when compute nodes query the scheduler to check for running/pending jobs this might fail and the instance is reported as idle, hence terminated after the configured scaledown_idletime (10 mins by default). SOLUTION: choose a more performant instance type for the master node.
- you are running a fairly long post install script which is taking approx the same time as the configured scaledown_idletime. The compute node is getting attached to the scheduler and at the same time the nodewatcher is terminating it. This explains also why some instances are failing the EC2_INSTANCE_TERMINATE with `Instance not found in database` error. 
SOLUTION: increase scaledown_idletime or upgrade to 2.4.0 where the idle time timer is only started once post_install script is completed

Unfortunately the only way to retrieve logs from a terminated instance is to integrate with CloudWatch as described here: https://github.com/aws/aws-parallelcluster/wiki/CloudWatch-Logs. We do plan to have CloudWatch logs natively supported by ParallelCluster in the future but for the moment you have to configure it manually.

As part of the release 2.4.0 some additional checks have been put in place in the nodewatcher and in case the node is terminating itself, due to the scheduler being unresponsive, the logs are dumped to the shared home dir (`/home/logs`). Also in case a failing node is terminated and a job is running on it, the job is automatically terminated or rescheduled according to the submission options. In particular these are the changes to the SGE global scheduler config for version 2.4.0:
- max_unheard 00:03:00: allows a faster reaction in case of faulty nodes
- reschedule_unknown 00:00:30: enables rescheduling of jobs running on failing nodes
- qmaster_params ENABLE_FORCED_QDEL_IF_UNKNOWN: forces job deletion on unresponsive nodes
- qmaster_params ENABLE_RESCHEDULE_KILL: forces rescheduling or killing of jobs running on failing nodes

To summarise I would recommend you to:
- Choose a bigger master instance type such as c5.xlarge
- Use pcluster 2.4.0 which offers a better overall experience and error handling especially with spot instances

**Comment by msherman13** [comment]
@demartinofra

thank you for the detailed response; very informative. I suspect I have found the root cause of this issue, you were definitely on the right track when you mentioned something about the post-install script. Looks like the script was intermittently failing. I fixed that and the issue seems to be resolved.

I will also take your advice and upgrade my master node. Unfortunately, regarding upgrading ParallelCluster itself, I am blocked because of the below issue:

https://github.com/aws/aws-parallelcluster/issues/1142

**Comment by demartinofra** [comment]
I just answered on the other issue as well. Resolving this. Feel free to reopen if necessary.
